### PR TITLE
Add intercept parameter to Polynomial kernel

### PR DESCRIPTION
**Pull request by peterroelants**
Add intercept parameter to the Polynomial kernel as was suggested as a future addition in https://github.com/tensorflow/probability/pull/194

### CODE REVIEW

File: tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 from __future__ import division
 from __future__ import print_function
 
+
+from functools import reduce
```
Comment (csuter) [inline tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py]: as a style rule we import only modules (or submodules), but not individual exported items from modules. please import functools here and use functools.reduce below

File: tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py [hunk 2]
```diff
@@ -142,26 +161,32 @@ def slope_variance(self):
...
-            _maybe_shape_dynamic(self.bias_variance)),
-        _maybe_shape_dynamic(self.exponent))
+    return reduce(
+      tf.broadcast_dynamic_shape,
+      map(_maybe_shape_dynamic, [self.slope_variance, self.bias_variance,
+                                 self.intercept, self.exponent]))
 
   def _apply(self, x1, x2, param_expansion_ndims=0):
+    if self.intercept is not None:
+      x1 = x1 - self.intercept
```
Comment (csuter) [inline tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py]: nit: maybe just inline these below; feels weird to re-interpret the meanings of `x1` and `x2`

File: tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py [hunk 3]
```diff
@@ -82,17 +95,18 @@ def testNoneShapes(self):
     self.assertEqual([2, 3, 2], k.batch_shape.as_list())
 
   @parameterized.named_parameters(
-      ('Shape [] kernel', 2., 2., 2., []),
-      ('Shape [1] kernel', [2.], [2.], [2.], [1]),
-      ('Shape [2] kernel', [1., 2.], [1., 2.], [1., 2.], [2]),
+      ('Shape [] kernel', 2., 2., 2., 2., []),
+      ('Shape [1] kernel', [2.], [2.],[2.],  [2.], [1]),
```
Comment (csuter) [inline tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py]: missing a space after comma

File: tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py [hunk 4]
```diff
@@ -82,17 +95,18 @@ def testNoneShapes(self):
     self.assertEqual([2, 3, 2], k.batch_shape.as_list())
 
   @parameterized.named_parameters(
-      ('Shape [] kernel', 2., 2., 2., []),
-      ('Shape [1] kernel', [2.], [2.], [2.], [1]),
-      ('Shape [2] kernel', [1., 2.], [1., 2.], [1., 2.], [2]),
+      ('Shape [] kernel', 2., 2., 2., 2., []),
```
Comment (csuter) [inline tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py]: these expressions are a bit hard to parse mentally. i think these parameter entries can be kwargs -- could you add them (matching the method kwargs), which i think will also force one arg per line and be a bit more readable?

File: tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py [hunk 5]
```diff
@@ -171,17 +189,20 @@ def testShapesAreCorrectBroadcast(self):
...
         slope_variance=slope_variance,
+        intercept = intercept,
         exponent=exponent
     )
     x = np.random.uniform(-1, 1, size=[5, 3]).astype(np.float32)
     y = np.random.uniform(-1, 1, size=[4, 3]).astype(np.float32)
     self.assertAllClose(
         self.evaluate(k.matrix(x, y)),
-        bias_variance**2 + slope_variance**2 * (x.dot(y.T))**exponent
+        (bias_variance**2 + slope_variance**2 *
```
Comment (csuter) [inline tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py]: nit: spaces around operators (including exponentiation)

**Comment by peterroelants** [comment]
@csuter I amended and pushed the changes you recommended.

**Comment by csuter** [comment]
Apologies, @peterroelants -- can you fix the enclosed lint errors?

```
Wrong hanging indentation (add 2 spaces).
      bias_variance, slope_variance, intercept, exponent]):
      ^ |
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py:112
Wrong hanging indentation (add 2 spaces).
      tf.broadcast_static_shape,
      ^ |
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py:162
Wrong hanging indentation (add 2 spaces).
      map(_maybe_shape_static, [self.slope_variance, self.bias_variance,
      ^ |
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py:163
Wrong hanging indentation (add 2 spaces).
      tf.broadcast_dynamic_shape,
      ^ |
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py:168
Wrong hanging indentation (add 2 spaces).
      map(_maybe_shape_dynamic, [self.slope_variance, self.bias_variance,
      ^ |
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py:169
Wrong hanging indentation (add 1 space).
       intercept=-1.,
       ^|
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py:64
Wrong hanging indentation (add 1 space).
       validate_args=True)
       ^|
 [bad-continuation]
go/gpylint-faq#bad-continuation
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py:65
No space allowed around keyword argument assignment
        intercept = intercept,
                  ^
 [bad-whitespace]
go/gpylint-faq#bad-whitespace
        //depot/google3/third_party/py/tensorflow_probability/python/positive_semidefinite_kernels/polynomial_test.py:239
Analyzed 2 file(s). 8 new error(s) and 0 non-actionable issue(s) found.
```

**Comment by peterroelants** [comment]
@csuter I should've fixed the linter warnings, amended the commit and pushed the changes.

**Comment by csuter** [comment]
Hi @peterroelants sorry for the huge lag on this, but...happy new year!

We have one more request, which came up when we started pulling this in. Elsewhere in TFP we refer to parameters like this either as `"loc"` (as in location/scale family distributions like `Normal`) or `"shift"`. We'd prefer not to introduce yet a 3rd noun for a similar concept, so we'd like to use one of those two terms. I kinda favor "shift" here. Any thoughts? Do you mind making the change, and then I can finalize the pulling in of the PR?

Thanks again!

**Comment by peterroelants** [comment]
@csuter Happy new year.

I updated the PR to use `shift` instead of `intercept`. I rebased, amended, and pushed the changes.

I was however not able to test it locally. After updating Bazel and running the tests I got a `ERROR: missing input file '//tensorflow_probability:leakr_badwords.dic'`. I currently won't have time to dig into this seemingly unrelated error myself. I noticed that a CI build process was added to the repository, so hopefully, this returns a positive result.

**Comment by csuter** [comment]
Thanks, @peterroelants and sorry for the build confusion. I'm working on some fixes right now.

In the mean time, there is a valid lint error (super unobvious -- also working on fixing that!). On line 140, there's an extra 2-space indent. I think the lines continuing the parenthetical are also indented an extra 2, so the whole chunk can be pulled back 2 spaces and then it should be good to go.

**Comment by peterroelants** [comment]
@csuter Updated to fix the linter error. Checks seem to have passed this time.
